File: ferric_continuum/foundation/parameter_passing_demo.cc
#include "absl/log/globals.h"
#include "absl/log/initialize.h"
#include "absl/log/log.h"

#include <utility>

#include "parameter_passing.hh"

namespace ff = ferric::foundation;

int main() {
  // Initialize Abseil logging
  absl::InitializeLog();
  absl::SetStderrThreshold(absl::LogSeverityAtLeast::kInfo);

  LOG(INFO) << "=== Parameter Passing Demo ===";

  ff::Rectangle rect{10.0, 5.0};
  LOG(INFO) << "Original: " << rect.to_string();

  // 1. Pass by const reference - read-only, no copy
  LOG(INFO) << "1. Pass by const reference:";
  double area = ff::compute_area_by_const_ref(rect);
  LOG(INFO) << "   Area: " << area;
  LOG(INFO) << "   Original unchanged: " << rect.to_string();
  LOG(INFO) << "   Use case: Efficient read-only access";

  // 2. Pass by value - creates copy
  LOG(INFO) << "2. Pass by value (creates copy):";
  area = ff::compute_area_by_value(rect);
  LOG(INFO) << "   Computed area (width doubled internally): " << area;
  LOG(INFO) << "   Original unchanged: " << rect.to_string();
  LOG(INFO) << "   Use case: Need local modifications without affecting caller";

  // 3. Pass by reference - modifies original
  LOG(INFO) << "3. Pass by non-const reference:";
  LOG(INFO) << "   Before: " << rect.to_string();
  ff::scale_by_reference(rect, 2.0);
  LOG(INFO) << "   After:  " << rect.to_string() << " (modified!)";
  LOG(INFO) << "   Use case: Need to modify the original object";

  // Reset
  rect = {10.0, 5.0};

  // 4. Pass by pointer - nullable, can modify
  LOG(INFO) << "4. Pass by pointer:";
  LOG(INFO) << "   Before: " << rect.to_string();
  ff::scale_by_pointer(&rect, 2.0);
  LOG(INFO) << "   After:  " << rect.to_string() << " (modified!)";
  ff::scale_by_pointer(nullptr, 2.0);  // Safe - nullptr check inside
  LOG(INFO) << "   nullptr is safe (checked inside function)";
  LOG(INFO) << "   Use case: Optional parameters or C-style APIs";

  // 5. Pass by rvalue reference - consumes temporary
  LOG(INFO) << "5. Pass by rvalue reference:";
  ff::Rectangle temp{5.0, 3.0};
  LOG(INFO) << "   Temp before: " << temp.to_string();
  ff::Rectangle result = ff::transform_by_rvalue(std::move(temp), 3.0);
  LOG(INFO) << "   Result: " << result.to_string();
  LOG(INFO) << "   Use case: Efficiently consuming temporary objects";

  LOG(INFO) << "Guidelines:";
  LOG(INFO) << "- const& : Default choice for read-only access";
  LOG(INFO) << "- value  : Small objects or when you need a local copy";
  LOG(INFO) << "- &      : When you need to modify the original";
  LOG(INFO) << "- *      : For optional parameters or C APIs";
  LOG(INFO) << "- &&     : For move semantics and consuming temporaries";

  return 0;
}

File: ferric_continuum/foundation/parameter_passing_demo.rs
use parameter_passing::{
    borrow_large_rect, compute_area_by_ref, compute_area_by_value, consume_large_rect,
    scale_by_mut_ref, transform_by_value, LargeRectangle, Rectangle,
};
use tracing::{info, Level};
use tracing_subscriber;

fn main() {
    // Initialize logging
    tracing_subscriber::fmt().with_max_level(Level::INFO).init();

    info!("=== Parameter Passing Demo (Rust) ===");

    let rect = Rectangle::new(10.0, 5.0);
    info!(original = rect.to_string(), "Starting rectangle");

    // 1. Pass by immutable reference - most common
    info!("1. Pass by immutable reference:");
    let area = compute_area_by_ref(&rect);
    info!(area, "Area calculated");
    info!(rect = rect.to_string(), "Original unchanged");
    info!("Use case: Read-only access (most common in Rust)");

    // 2. Pass by value - copies for Copy types
    info!("2. Pass by value (Copy trait):");
    let area = compute_area_by_value(rect);
    info!(area, "Computed area (width doubled internally)");
    info!(rect = rect.to_string(), "Original unchanged (Copy trait)");
    info!("Use case: Small types that implement Copy");

    // 3. Pass by mutable reference - modify in place
    info!("3. Pass by mutable reference:");
    let mut rect_mut = Rectangle::new(10.0, 5.0);
    info!(before = rect_mut.to_string(), "Before modification");
    scale_by_mut_ref(&mut rect_mut, 2.0);
    info!(after = rect_mut.to_string(), "After modification");
    info!("Use case: Modify the original value");

    // 4. Pass by value and transform
    info!("4. Pass by value and transform:");
    let rect2 = Rectangle::new(5.0, 3.0);
    info!(before = rect2.to_string(), "Before transform");
    let result = transform_by_value(rect2, 3.0);
    info!(result = result.to_string(), "After transform");
    info!("Use case: Transform and return new value");

    // 5. Move semantics with non-Copy types
    info!("5. Move semantics (non-Copy type):");
    let large = LargeRectangle::new(10.0, 5.0);

    // Borrow - still usable after
    info!("Borrowing...");
    let area = borrow_large_rect(&large);
    info!(area, "Area from borrow");
    info!("large is still accessible after borrow");

    // Consume - NOT usable after
    info!("Consuming (move)...");
    let area = consume_large_rect(large);
    info!(area, "Area from consume");
    info!("large is NO LONGER accessible (moved)");

    info!("Key Differences from C++:");
    info!("- Rust: Borrow by default (&), move is explicit or automatic");
    info!("- C++: Copy by default, move needs std::move()");
    info!("- Rust enforces borrowing rules at compile time");
    info!("- No null pointers in safe Rust (use Option<&T> instead)");

    info!("Rust Guidelines:");
    info!("- &T      : Default for read-only (most common)");
    info!("- &mut T  : When you need to modify");
    info!("- T       : Takes ownership (move for non-Copy, copy for Copy)");
    info!("- No need for pointer/reference choices - compiler enforces safety!");
}
